Add changelog file (#1962)
Nemocas/AbstractAlgebra.jl: CHANGELOG.md

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/), and this project
tries to adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

> ## [Unreleased]

Reviewer: Thanks @aaruni96 

Can you please also add the `release notes: XYZ` labels to the AA and Nemo repositories? And then perhaps backfill a few recent releases. e.g. here going back to 0.44.0 shouldn't be too much, it's just a handful of changes (I think we should also consider backfilling at least `0.X.0` releases to a certain point, listing just the breaking changes...)

To be clear, backfilling might involve identifying the merged commits (via `git log`) and taking the PR titles. If those are bad, we will see it here and then we can request the authors to suggest better ones (in many cases that's me, and in others I'll still be able to suggest something, too)

Author: > Can you please also add the release notes: XYZ labels to the AA and Nemo repositories?

Done ( at least, I think I did what you meant for me to do). Can you check?

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/), and this project
tries to adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Started keeping a changelog!

## [0.44.3] - 2025-01-17

### Added

- Add conformance tests for `MPolyRing`, and fix bugs in `^` and `is_unit` for `MPoly` [PR
  #1950](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1950)

- Add optional `ignore_kwargs` argument to `@attr` macro [PR
  #1958](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1958)

### Changed

- Don't require `base_ring` in ring conformance tests [issue
  #1944](https://github.com/Nemocas/AbstractAlgebra.jl/issues/1944) [PR
  #1946](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1946)

- Fix promotion for matrix-scalar operations [PR
  #1949](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1949)

> - Replace uses of elem_type(coefficient_ring(R)) by equivalent compile-time constructs [PR
>   #1956](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1956)

Reviewer: ```suggestion
- Some internal enhancements
```

## [0.44.2] - 2024-12-22

### Changed

- Make `is_exact_type` and `is_domain_type` more convenient by
  having them accept rings and ring elements (not just types) [PR
  #1942](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1942)

### Fixed

> - Fix `is_unit` for univariate and multivariate polynomials over `Z/nZ` [issue
>   #11](https://github.com/Nemocas/AbstractAlgebra.jl/issues/11) [PR
>   #1933](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1933)

Reviewer 2: IMO we don't need to mention closed issues in most cases, since if they are related to a change, we can instead link the PR which usually refers to the issue itself. This would simplify as follows:
```suggestion
- Fix `is_unit` for univariate and multivariate polynomials over `Z/nZ` [#1933](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1933)
```

Author: The idea was that people waiting for a specific issue to be resolved could ctrl+f the changelog file. I don't know if we have many such users, so we could also go with your suggestion. I am fine either way.

Reviewer: I don't mind listing issue numbers (it doesn't hurt) but for the automation they are not trivial to get in general. So I don't to require them.

It may be possible to improve the script to at least see if a PR has linked issues (I am guessing GH API allows us to figure those out), then we could reference those. But that's for the down the line

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/), and this project
tries to adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Started keeping a changelog!

## [0.44.3] - 2025-01-17

### Added

> - Add conformance tests for MPolyRing, and fix bugs in ^ and is_unit for MPoly [PR

Reviewer: ```suggestion
- Add conformance tests for `MPolyRing`, and fix bugs in `^` and `is_unit` for `MPoly` [PR
```

  #1950](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1950)

> - Add ignore_kwargs to @attr macro [PR

Reviewer 2: This will get deprecated again in https://github.com/Nemocas/AbstractAlgebra.jl/pull/1966, so no real need to add it at all.

Reviewer: I would disagree, to the contrary, someone may have stumbled over the new feature and used it, and then it is suddenly gone. So instead of just never mentioning it, if the changelog has an entry where it gets added and then another where it gets deprecated, that helps users to adjust.

(It may not be very important here because it was so short lived but I don't see why to deviate from the principle here)

But this should be formatted with `code blocks`:
```suggestion
- Add optional `ignore_kwargs` argument to `@attr` macro [PR
```

  #1958](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1958)

### Changed

> - Don't require base_ring in ring conformance tests [issue

Reviewer: ```suggestion
- Don't require `base_ring` in ring conformance tests [issue
```

  #1944](https://github.com/Nemocas/AbstractAlgebra.jl/issues/1944) [PR
  #1946](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1946)

> - Fixed promotion for matrix-scalar operations [PR

Reviewer: ```suggestion
- Fix promotion for matrix-scalar operations [PR
```

  #1949](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1949)

- Replace uses of elem_type(coefficient_ring(R)) by equivalent compile-time constructs [PR

>   #1956](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1956)

Reviewer 2: I wouldn't add this. it's internal optimizations only

Reviewer: Agreed.

I tend to have a generic entry like `- various internal enhancements` or `- various janitorial changes` etc., as and if appropriate.

## [0.44.2] - 2024-12-22

### Changed

> - Re-enable conformance test for residue ring over Z[x] [issue
>   #994](https://github.com/Nemocas/AbstractAlgebra.jl/issues/994) [PR
>   #1938](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1938)
> 

Reviewer: This is just adding an internal test
```suggestion
```

> - Make is_exact_type and is_domain_type more convenient [issue

Reviewer: ```suggestion
- Make `is_exact_type` and `is_domain_type` more convenient by
  having them accept rings and ring elements (not just types) [PR
```

  #1942](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1942)

### Fixed

> - Fix isunit for univariate and multivariate polynomials over Z/nZ [issue

Reviewer: ```suggestion
- Fix `is_unit` for univariate and multivariate polynomials over `Z/nZ` [issue
```
